### PR TITLE
feat: assembler parser module

### PR DESCRIPTION
**Pull request by henok-tesfaye**
Make progress on #19 

**What kind of change does this PR introduce?**
It is a feature, which adds an assembler parser module

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
Nothing

### CODE REVIEW

File: src/abstractions/software/assembler/parser/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ * @returns {string} label name
+ */
+export const getLabelName = instruction => {
+  if (!isLabelInstruction(instruction)) return ''
+  return LABEL_INSTRUCTION.exec(instruction)[FIRST_MATCH]
+}
+
+/**
+ * @param {string} instruction instruction
+ * @returns {string} adress number
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/index.js:71]:
```suggestion
 * @returns {string} address symbol or number
```

File: src/abstractions/software/assembler/parser/utils/index.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+// TODO
+/**
+ * @param {string} instruction instruction
+ * @returns {boolean} checks if it is a C(compute) instruction
+ */
+export const isComputeInstruction = instruction => {}
+
+/**
+ * @param {string} instruction instruction
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/index.js:61]:
```suggestion
 * @param {string} instruction instruction of type L_COMMAND
```

File: src/abstractions/software/assembler/parser/utils/index.test.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    const expectedJump = ''
+    const actualLJump = dest(aInstruction)
+    const actualAJump = dest(lInstruction)
+
+    expect(actualAJump).toBe(expectedJump)
+    expect(actualLJump).toBe(expectedJump)
+  })
+})
+
+const getAssmCode = () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/index.test.js:215]:
```suggestion
const getAssemblyTestCode = () => {
```

File: src/abstractions/software/assembler/parser/utils/index.test.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+describe('getLabelName function', () => {
+  it('returns the label name if it\'s a label instruction', () => {
+    const instruction = '(Xxx.xx)'
+    const expectedLabelName = 'Xxx.xx'
+    const actualLabelName = getLabelName(instruction)
+    expect(actualLabelName).toBe(expectedLabelName)
+  })
+
+  it(`returns the label name even if there is multiple spaces betweeen open
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/index.test.js:109]:
```suggestion
  it(`returns the label name even if there are multiple spaces between open
```

File: src/abstractions/software/assembler/parser/utils/index.test.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    const actualLDestination = dest(aInstruction)
+    const actualADestination = dest(lInstruction)
+
+    expect(actualADestination).toBe(expectedDestination)
+    expect(actualLDestination).toBe(expectedDestination)
+  })
+})
+
+describe('comp function', () => {
+  it('returns the correct compute if instruction is C Type and has both destination and jump', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/index.test.js:162]:
```suggestion
  it('returns the correct computation if instruction is C Type and has both destination and jump', () => {
```

File: src/abstractions/software/assembler/parser/utils/index.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  return LABEL_INSTRUCTION.exec(instruction)[FIRST_MATCH]
+}
+
+/**
+ * @param {string} instruction instruction
+ * @returns {string} adress number
+ */
+export const getAddressName = instruction => {
+  if (!isAddressInstruction(instruction)) return ''
+  return ADRESS_INSTRUCTION.exec(instruction)[FIRST_MATCH]
```
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/parser/utils/index.js:75]: Actually `FIRST_MATCH` is defined to make it clear what `1` is in the regex match.

File: src/abstractions/software/assembler/parser/utils/regex.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export const REMOVE_COMMENT = /\/\/[^\n]+/g
+export const LABEL_INSTRUCTION = /\(\s*([A-Za-z_][_A-Za-z.$\d]*)\s*\)/
+export const ADRESS_INSTRUCTION = /^@([A-Za-z_\d][_A-Za-z.$\d]*)\b$/
```
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/parser/utils/regex.js:3]: `/@([A-Za-z_\d][_A-Za-z.$\d]*)\b$/.test('a@13') // true`
`/^@([A-Za-z_\d][_A-Za-z.$\d]*)\b$/.test('a@13') // false`

We need the word boundary, `^`, to check if it starts with `@`.

File: src/abstractions/software/assembler/parser/utils/regex.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export const REMOVE_COMMENT = /\/\/[^\n]+/g
+export const LABEL_INSTRUCTION = /\(\s*([A-Za-z_][_A-Za-z.$\d]*)\s*\)/
```
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/parser/utils/regex.js:2]: > Constants and Symbols Constants must be non-negative and are written in decimal notation. A user-defined symbol can be any sequence of letters, digits, underscore (_), dot (.), dollar sign ($), and colon (:) that does not begin with a digit.

I didn't know that, `:` is allowed symbol.

[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/parser/utils/regex.js:2]: `export const ADRESS_INSTR_REGEX = new RegExp(String.raw`@([${symbolChars}\d]*$)`)` the last `*` should be `+`.

File: src/abstractions/software/assembler/parser/utils/regex.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/parser/utils/regex.js:3]: I added a test for `regex.js`

File: src/abstractions/software/assembler/parser/utils/regex.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+const symbolChars = 'A-Za-z_.$:'
+export const REMOVE_COMMENT = /\/\/[^\n]*/g
```
Comment (mezzzi) [inline src/abstractions/software/assembler/parser/utils/regex.js:2]:
```suggestion
export const COMMENT_REGEX = /\/\/[^\n]*/g
```

**Pull request merged**